### PR TITLE
GSoC branch lang controller addition

### LINKED ISSUE #798: Add Multilingual Support to the Talawa App
### Parent Issue

**Is your feature request related to a problem? Please describe.**
Talawa Needs to have multilingual support to make it internationally acceptable.

**Describe the solution you'd like**
We need a way for users to easily select their language of choice which would immediately and permanently change the text to the new language until the language selection is changed.

### Child Issues:

The following are child issues for this parent:

1. #799 
2. #808
2. #800
3. #801
3. #802
4. #803
5. #804
6. #805
7. #806
8. #807

### LINKED ISSUE #800: Refactor Hard Coded strings
**Is your feature request related to a problem? Please describe.**
This is a child issue of #798

**Describe the solution you'd like**
1. Refactor all widgets in the Talawa app which have hardcoded strings.
2. Update the Talawa GitHub action workflow to automatically reject any hard coded strings outside of the language translation files.

### LINKED ISSUE #801: Language Selection
**Is your feature request related to a problem? Please describe.**
This is a child issue of #798

**Describe the solution you'd like**
1. Develop language selection page & create change language widgets on certain specific pages to allow users to easily switch language within the app.
2. We prefer to have language selection upon registration. Future changes would occur in the user settings screen

### PR DESCRIPTION
**Pull request by sumitra19jha**
**What kind of change does this PR introduce?**
Ability to update/change language through controller

**Issue Number:**

Fixes #798 
Fixes #800 before auth
Fixes #801 

**Did you add tests for your changes?**
No, Seprate issue exists for that

**Snapshots/Videos:**

https://user-images.githubusercontent.com/35868598/122646387-7bc9d580-d13c-11eb-9f0b-3b2ffe1294af.mp4

**If relevant, did you update the documentation?**
Not needed here

**Summary**
Develop branch language update 

**Does this PR introduce a breaking change?**
No

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**

### CODE REVIEW

**Comment by DeltaHarbinger** [comment]
Hey @sumitra19jha, some of the tests are failing on this PR. Is there something that needs to be changed with your code or with the pipelines?

**Comment by sumitra19jha** [comment]
Yeah, I am working bon them. It is related to a notifier problem.

**Comment by DeltaHarbinger** [comment]
Alright, thanks for the update

**Comment by CyberWake** [comment]
@sumitra19jha can you please look in to the thing like in verify text in some languages goes to 2 line plus the greeting text in english and french come as intended but in Chinese and hindi it doesn't

**Comment by sumitra19jha** [comment]
@CyberWake Ok. Working on improving that

**Comment by sumitra19jha** [comment]
@DeltaHarbinger During the language conversion the context is very important, which is given by the full sentence. In the case of Hindi and Chinese the meaning actually gets affected if we ended up breaking it like english. So full sentence is taken as a reference. 

The problem here is also the widget. While designing the UI, we focused on single language and then widgets are implemented accordingly (for the case of greeting the use of Rich Text). So the JSON are made like that.

### For English
Screenshot 2021-06-22 at 9 49 41 PM

## For Hindi
Screenshot 2021-06-22 at 9 49 56 PM

## For Spanish
Screenshot 2021-06-22 at 9 50 09 PM

## For Chinese
Screenshot 2021-06-22 at 9 50 17 PM

**Comment by sumitra19jha** [comment]
There is one issue in this PR which is causing error during signup so I have marked it as a draft. I will fix this as soon as possible.

**Comment by sumitra19jha** [comment]
@DeltaHarbinger @CyberWake @dellyjm @Sagar2366 @rutvik11062000 @utkarshshendge Please Review the PR

https://user-images.githubusercontent.com/35868598/122994103-b5614180-d3c5-11eb-8843-86142c6f0ef4.mp4

**Comment by sumitra19jha** [comment]
The PR is ready to be merged

**Comment by CyberWake** [comment]
> @DeltaHarbinger @CyberWake @dellyjm @Sagar2366 @rutvik11062000 @utkarshshendge Please Review the PR
> 
>  WhatsApp.Video.2021-06-23.at.1.51.22.AM.mp4

I feel like having the languages on the language page itself in the same locale would we much better, this is just the input what other social/management apps do.

**Comment by CyberWake** [comment]
Please look into the test as well they are failing.

**Comment by sumitra19jha** [comment]
@CyberWake Travis CI is passing

**Comment by CyberWake** [comment]
> @CyberWake Travis CI is passing. Can you point out which test cases are failing?

Try running locally you'll get to know if it's not clear with the logs over here.

**Comment by sumitra19jha** [comment]
@CyberWake Can you attach the snapshot of which test failed locally?

**Comment by sumitra19jha** [comment]
@CyberWake Locally everything is working fine on my system!!!

**Comment by CyberWake** [comment]
> @CyberWake Locally everything is working fine on my system!!!

did you really check it?

**Comment by CyberWake** [comment]
> @CyberWake Locally everything is working fine on my system!!!

did you check it? like from logs itself I see the test are failing in ```select_language_page_test.dart``` file. And error would also be thrown irrespective of machine its running on please cross check again.

**Comment by sumitra19jha** [comment]
@CyberWake Please share your error log screenshot. I am not able to see any on my local system.

This is the screenshot of `select_language_page_test.dart test` on my local system.
Screenshot 2021-06-24 at 9 00 01 AM

**Comment by CyberWake** [comment]
> @CyberWake Please share your error log screenshot. I am not able to see any on my local system.
> 
> This is the screenshot of `select_language_page_test.dart test` on my local system.
> Screenshot 2021-06-24 at 9 00 01 AM

please look into the line 5 widgets with key and for more please run the test in android studio if possible.

**Comment by sumitra19jha** [comment]
@CyberWake There is no issue with line 5. This is appearing because of print statement, as we have added same key to more than 1 items. If you are facing any issue, Please message me on slack with screenshot and we will work on that.

https://user-images.githubusercontent.com/35868598/123221351-7d085300-d4ec-11eb-91dc-1646c8df20e5.mov

.

**Comment by CyberWake** [comment]
> @CyberWake There is no issue with line 5. This is appearing because of print statement, as we have added same key to more than 1 items. If you are facing any issue, Please message me on slack with screenshot and we will work on that.
> 
>  screen-recording-2021-06-24-at-125726-pm_iUMuOMvi.mov 
> .

please run the test in android studio the full statement is not shown here i would recommend you to use android studio to run test there you get better logs. and the exact log could itself be seen from the workflow adding for convenience
```The following TestFailure object was thrown running a test:
  Expected: BoxDecoration:<BoxDecoration(color: Color(0x00000000))>
  Actual: BoxDecoration:<BoxDecoration(color: Color(0x26c4c4c4))>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///home/runner/work/talawa/talawa/test/widget_tests/select_language_page_test.dart:235:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
...

This was caught by the test expectation on the following line:
  file:///home/runner/work/talawa/talawa/test/widget_tests/select_language_page_test.dart line 235
The test description was:
  Testing to change language items
```

**Comment by sumitra19jha** [comment]
@CyberWake This is happening because the `async` request on your system is running slow. I will fix this in next commit.

I am inclined to work for both frontend and backend so I am more comfortable with VS Code then Android Studio. For testing purposes it works very fine.

**Comment by CyberWake** [comment]
> @CyberWake This is happening because the `async` request on your system is running slow. I will fix this in next commit.
> 
> I work on both frontend and backend so I am more comfortable with VS Code then Android Studio. For both development and testing purposes it works very fine.

can you please run the test in android studio it is not something dependent on machine

**Comment by sumitra19jha** [comment]
@DeltaHarbinger @CyberWake @dellyjm @rutvik11062000 @utkarshshendge @Sagar2366 
Please review the PR its ready

**Comment by CyberWake** [comment]
> @DeltaHarbinger @CyberWake @dellyjm @rutvik11062000 @utkarshshendge @Sagar2366
> Please review the PR its ready

please look up on the review comments

**Comment by sumitra19jha** [comment]
What review comments? @CyberWake

**Comment by sumitra19jha** [comment]
Is there any issue in these changes? @CyberWake

**Review by CyberWake — CHANGES_REQUESTED** [review]

File: lib/views/pre_auth_screens/select_language.dart [hunk 1]
```diff
@@ -39,16 +32,18 @@ class _SelectLanguageState extends State<SelectLanguage> {
...
+                key: const Key('Select Language'),
               ),
             ),
             SizedBox(
               height: SizeConfig.screenHeight! * 0.018,
             ),
-            const CupertinoSearchTextField(
-              key: Key('SearchField'),
-            ),
+            // const CupertinoSearchTextField(
```
[the remark this answers is not in the document]
Comment (sumitra19jha) [reply lib/views/pre_auth_screens/select_language.dart:44]: This is not required right now. This is the reason I removed it, moreover why are you using CupertinoSearchTextField when major use base is for android

**Comment by CyberWake** [comment]
sorry i forgot to submit my review 😥

File: lib/views/pre_auth_screens/select_language.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (sumitra19jha) [reply lib/views/pre_auth_screens/select_language.dart:44]: If more than 10 languages will be added then I will make sure to implement search (but still I wouldn't recommend to use CupertinoSearchTextField in that case too)

File: lib/router.dart [hunk 2]
```diff
@@ -31,36 +31,36 @@ Route<dynamic> generateRoute(RouteSettings settings) {
       return MaterialPageRoute(
           builder: (context) => const SplashScreen(key: Key('SplashScreen')));
     case Routes.languageSelectionRoute:
-      final int selectedLangId = int.parse(settings.arguments!.toString());
```
[the remark this answers is not in the document]
Comment (sumitra19jha) [reply lib/router.dart:34]: it will not work due to lack of json files, for languages currently. But, this will be done in July though.

**Comment by sumitra19jha** [comment]
@CyberWake @rutvik11062000 @utkarshshendge @DMills27 @palisadoes Please review

**Review by palisadoes — APPROVED** [review]

**Pull request merged**